File: app/src/main/java/com/app/littlechat/CommonUtilities.java
package com.app.littlechat;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.Point;
import android.graphics.drawable.ColorDrawable;
import android.media.ExifInterface;
import android.util.Log;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.RecyclerView;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.style.UnderlineSpan;
import android.view.Display;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;

public class CommonUtilities {

    private static Dialog dialog;


    public static void putString(Activity activity, String name, String value)
    {
        SharedPreferences preferences=activity.getSharedPreferences(BuildConfig.APPLICATION_ID,Context.MODE_PRIVATE);

        preferences.edit().putString(name,value).apply();
    }

    public static String getString(Activity activity, String name)
    {
        SharedPreferences preferences=activity.getSharedPreferences(BuildConfig.APPLICATION_ID,Context.MODE_PRIVATE);

        return  preferences.getString(name,"");
    }

    public static void clearPrefrences(Activity activity)
    {
        SharedPreferences preferences=activity.getSharedPreferences(BuildConfig.APPLICATION_ID,Context.MODE_PRIVATE);

        preferences.edit().clear().apply();
    }


    public static void showLogoutPopup(final Activity activity) {
        AlertDialog.Builder alert = new AlertDialog.Builder(activity);

        alert.setCancelable(true);

        alert.setMessage("Do you want to logout?");

        alert.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                clearPrefrences(activity);
                activity.startActivity(new Intent(activity,MainActivity.class).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK));
            }
        });

        alert.setNegativeButton("No", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                dialog.cancel();
            }
        });


        alert.show();
    }

    public static void showToast(Activity activity, String message) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show();
    }


    public final static boolean isValidEmail(CharSequence target) {
        return !TextUtils.isEmpty(target) && android.util.Patterns.EMAIL_ADDRESS.matcher(target).matches();
    }

    public static void underlineTextView(TextView textView, String text)
    {
        SpannableString content = new SpannableString(text);
        content.setSpan(new UnderlineSpan(), 0, content.length(), 0);
        textView.setText(content);
    }


    public static void showAlert(final Activity activity, String msg, final boolean isFinish) {
        AlertDialog.Builder alert = new AlertDialog.Builder(activity);

        alert.setCancelable(true);

        alert.setMessage(msg);

        alert.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                dialog.cancel();
                if(isFinish)
                    activity.finish();
            }
        });

        alert.show();
    }

    public static void setLayoutManager(RecyclerView view, RecyclerView.LayoutManager manager) {
        RecyclerView.LayoutManager mLayoutManager1 = manager;
        view.setLayoutManager(mLayoutManager1);
        view.setItemAnimator(new DefaultItemAnimator());
        view.setNestedScrollingEnabled(false);
    }



    public static void showProgressWheel(Context context) {
        hideProgressWheel();

        dialog = new Dialog(context);

        dialog.setCancelable(false);

        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

        dialog.setContentView(R.layout.layout_progress_wheel);

        dialog.getWindow().setBackgroundDrawable(new ColorDrawable(android.graphics.Color.TRANSPARENT));

        Display display = ((Activity) context).getWindowManager().getDefaultDisplay();

        Point size = new Point();

        display.getSize(size);

        int width = size.x;

        dialog.getWindow().setLayout(width - 60, ViewGroup.LayoutParams.WRAP_CONTENT);// set width of alert dialog box nad get width dynamically

        dialog.show();
    }
    public static void hideProgressWheel() {
        try {
            if (dialog != null)
                dialog.dismiss();
        } catch (Exception e) {

        }
    }




    public static String getResizedBitmap(String path, int maxSize, String fileName, Activity activity,boolean isCamera) {

        BitmapFactory.Options options = new BitmapFactory.Options();

        options.inPreferredConfig = Bitmap.Config.ARGB_8888;
        Bitmap image;
        if(isCamera)
            image = CommonUtilities.rotateBitmap(BitmapFactory.decodeFile(path, options),path);
        else
            image=BitmapFactory.decodeFile(path, options);

        int width = image.getWidth();
        int height = image.getHeight();

        float bitmapRatio = (float) width / (float) height;
        if (bitmapRatio > 1) {
            width = maxSize;
            height = (int) (width / bitmapRatio);
        } else {
            height = maxSize;
            width = (int) (height * bitmapRatio);
        }


        Bitmap saveBitmap = Bitmap.createScaledBitmap(image, width, height, true);

        File cacheDir = activity.getBaseContext().getCacheDir();

        File saveFile = new File(cacheDir, fileName);

        try {
            FileOutputStream out = new FileOutputStream(saveFile);
            saveBitmap.compress(Bitmap.CompressFormat.JPEG, 100, out);
            out.flush();
            out.close();

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return saveFile.getPath();
    }


    public static  Bitmap rotateBitmap(Bitmap realImage,String imagePath)
    {
        try {
            ExifInterface exif = new ExifInterface(imagePath);

            Log.d("EXIF value", exif.getAttribute(ExifInterface.TAG_ORIENTATION));
            if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equalsIgnoreCase("6")) {

                realImage = rotate(realImage, 90);
            } else if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equalsIgnoreCase("8")) {
                realImage = rotate(realImage, 270);
            } else if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equalsIgnoreCase("3")) {
                realImage = rotate(realImage, 180);
            }
            return realImage;
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        return realImage;
    }

    public static Bitmap rotate(Bitmap bitmap, int degree) {
        int w = bitmap.getWidth();
        int h = bitmap.getHeight();

        Matrix mtx = new Matrix();
        mtx.postRotate(degree);

        return Bitmap.createBitmap(bitmap, 0, 0, w, h, mtx, true);
    }
}

File: app/src/main/java/com/app/littlechat/utility/CommonUtilities.kt
//package com.app.littlechat.utility
//
//import android.app.Activity
//import android.app.AlertDialog
//import android.app.Dialog
//import android.content.Context
//import android.content.Intent
//import android.graphics.Bitmap
//import android.graphics.BitmapFactory
//import android.graphics.Color
//import android.graphics.Matrix
//import android.graphics.Point
//import android.graphics.drawable.ColorDrawable
//import android.media.ExifInterface
//import android.net.ConnectivityManager
//import android.text.SpannableString
//import android.text.TextUtils
//import android.text.style.UnderlineSpan
//import android.util.Log
//import android.view.ViewGroup
//import android.view.Window
//import android.widget.TextView
//import android.widget.Toast
//import androidx.recyclerview.widget.DefaultItemAnimator
//import androidx.recyclerview.widget.RecyclerView
//import com.app.littlechat.BuildConfig
//import com.app.littlechat.R
//import com.app.littlechat.data.model.User
//import java.io.File
//import java.io.FileNotFoundException
//import java.io.FileOutputStream
//import java.io.IOException
//
//
//class CommonUtilities {
//
//    companion object {
//
//
//        fun isNetworkConnected(context: Context) : Boolean{
//            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
//            val activeNetwork = cm.activeNetworkInfo
//            val isConnected = activeNetwork != null && activeNetwork.isConnectedOrConnecting
//            return isConnected
//        }
//
//        private var dialog: Dialog? = null
//
//
//        fun putString(activity: Context, name: String, value: String?) {
//            val preferences = activity.getSharedPreferences(BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE)
//
//            preferences.edit().putString(name, value).apply()
//        }
//
//        fun getString(activity: Context, name: String): String {
//            val preferences = activity.getSharedPreferences(BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE)
//
//            return preferences.getString(name, "")?:""
//        }
//
//
//        fun putToken(activity: Context, value: String?) {
//            val preferences = activity.getSharedPreferences(BuildConfig.APPLICATION_ID+"+token", Context.MODE_PRIVATE)
//
//            preferences.edit().putString("token", value).apply()
//        }
//
//        fun getToken(activity: Context): String {
//            val preferences = activity.getSharedPreferences(BuildConfig.APPLICATION_ID+"+token", Context.MODE_PRIVATE)
//
//            return preferences.getString("token", "")?:""
//        }
//
//        fun clearPrefrences(activity: Activity) {
//            val preferences = activity.getSharedPreferences(BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE)
//
//            preferences.edit().clear().apply()
//        }
//
//
//        fun showToast(activity: Activity, message: String) {
//            Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
//        }
//
//
//        fun isValidEmail(target: CharSequence): Boolean {
//            return !TextUtils.isEmpty(target) && android.util.Patterns.EMAIL_ADDRESS.matcher(target).matches()
//        }
//
//        fun underlineTextView(textView: TextView, text: String) {
//            val content = SpannableString(text)
//            content.setSpan(UnderlineSpan(), 0, content.length, 0)
//            textView.text = content
//        }
//
//
//        fun showAlert(activity: Activity, msg: String?, isFinish: Boolean, isCancelable : Boolean) {
//            val alert = AlertDialog.Builder(activity)
//
//            alert.setCancelable(isCancelable)
//
//            alert.setMessage(msg)
//
//            alert.setPositiveButton("Ok") { dialog, whichButton ->
//                dialog.cancel()
//                if (isFinish)
//                    activity.finish()
//            }
//
//            alert.show()
//        }
//
//        fun setLayoutManager(view: RecyclerView, manager: RecyclerView.LayoutManager) {
//            view.layoutManager = manager
//            view.itemAnimator = DefaultItemAnimator()
//            view.isNestedScrollingEnabled = false
//        }
//
//
//        fun hideProgressWheel() {
//            try {
//                if (dialog != null)
//                    dialog!!.dismiss()
//            } catch (e: Exception) {
//
//            }
//
//        }
//
//
//        fun getResizedBitmap(
//            path: String,
//            maxSize: Int,
//            fileName: String,
//            activity: Activity,
//            isCamera: Boolean
//        ): String {
//
//            val options = BitmapFactory.Options()
//
//            options.inPreferredConfig = Bitmap.Config.ARGB_8888
//            val image: Bitmap = if (isCamera)
//                rotateBitmap(BitmapFactory.decodeFile(path, options), path)
//            else
//                BitmapFactory.decodeFile(path, options)
//
//            var width = image.width
//            var height = image.height
//
//            val bitmapRatio = width.toFloat() / height.toFloat()
//            if (bitmapRatio > 1) {
//                width = maxSize
//                height = (width / bitmapRatio).toInt()
//            } else {
//                height = maxSize
//                width = (height * bitmapRatio).toInt()
//            }
//
//
//            val saveBitmap = Bitmap.createScaledBitmap(image, width, height, true)
//
//            val cacheDir = activity.baseContext.cacheDir
//
//            val saveFile = File(cacheDir, fileName)
//
//            try {
//                val out = FileOutputStream(saveFile)
//                saveBitmap.compress(Bitmap.CompressFormat.JPEG, 100, out)
//                out.flush()
//                out.close()
//
//            } catch (e: FileNotFoundException) {
//                e.printStackTrace()
//            } catch (e: IOException) {
//                e.printStackTrace()
//            }
//
//            return saveFile.path
//        }
//
//
//        fun rotateBitmap(realImage: Bitmap, imagePath: String): Bitmap {
//            var realImage = realImage
//            try {
//                val exif = ExifInterface(imagePath)
//
//                Log.d("EXIF value", exif.getAttribute(ExifInterface.TAG_ORIENTATION)?:"")
//                if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equals("6", ignoreCase = true)) {
//
//                    realImage = rotate(realImage, 90)
//                } else if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equals("8", ignoreCase = true)) {
//                    realImage = rotate(realImage, 270)
//                } else if (exif.getAttribute(ExifInterface.TAG_ORIENTATION).equals("3", ignoreCase = true)) {
//                    realImage = rotate(realImage, 180)
//                }
//                return realImage
//            } catch (e: Exception) {
//                e.printStackTrace()
//            }
//
//            return realImage
//        }
//
//        fun rotate(bitmap: Bitmap, degree: Int): Bitmap {
//            val w = bitmap.width
//            val h = bitmap.height
//
//            val mtx = Matrix()
//            mtx.postRotate(degree.toFloat())
//
//            return Bitmap.createBitmap(bitmap, 0, 0, w, h, mtx, true)
//        }
//
//        fun getUserData(activity: Activity): User {
//            return User(
//                getString(activity, Constants.SENDER_ID),
//                getString(activity, Constants.NAME),
//                getString(activity, Constants.EMAIL),
//                getString(activity, Constants.PHONE_NUMBER),
//                getString(activity, Constants.IMAGE),
//                getString(activity, Constants.STATUS)
//            )
//        }
//
//    }
//
//
//}